### DISCUSSION TITLE
How to use Rook cephfs erasure coding

### DISCUSSION THREAD

**Discussion by rich0** [rook/rook#12514] [Q&A] [answered]
I'm trying to experiment with cephfs and erasure coding, but there are very few Rook docs on the topic, and what exists doesn't make clear how to actually make use of it.  I'm having no issues getting replication filesystems to work.

The docs say that you can't actually create an erasure coding data pool only for a filesystem, but you need to create a default replicated pool and then add a secondary EC pool.  Indeed if I try to only create an EC data pool the operator log shows an error (22) when trying to create the filesystem.

If I go ahead and follow the example and create a size=1 replicated pool and a 2,2 EC pool, the filesystem is created, and I can mount it, but if I run "getfattr -n ceph.dir.layout ." in the root of the filesystem it shows that data will be stored in the replicated pool.  If I try to use setfattr to point it to the EC pool it gives me an invalid argument error (or Permission denied if I don't have the p capability set).

How do I go about actually creating a filesystem that uses EC to store data?  Also, can I just manually delete the replicated pool?  I'd prefer to not have more PGs floating around than needed.

I am getting the impression that not many are actually using EC with cephfs+Rook.  I'm interested in how it will perform as size=3 is obviously wasteful in terms of space use, and size=2 makes it difficult to have high availability safely.

**Comment by rich0** [comment]
Ok, I've made a bit of progress here.  I realized I'm using the wrong syntax on setfattr.

setfattr -n ceph.dir.layout.pool -v <pool> works fine.

It still seems like a bit of a kludge that rook forces you to create a replicated pool on an EC filesystem, but I can just work around that most likely by setting the pool at the root before loading it with data.  I can have an empty replicated pool with a low size+pg_num and that shouldn't really consume many resources.

**Comment by travisn** [comment, marked as answer]
@rich0 Did you create the storage class with the `pool` set to the EC pool name? See [this example](https://github.com/rook/rook/blob/master/deploy/examples/csi/cephfs/storageclass-ec.yaml#L19). Then it's expected the data would be stored in the EC pool. 
The replicated pool is still used for the filesystem metadata, so it's recommended to keep replica 3, otherwise you could lose the filesystem if replica is 1.

**Reply by rich0** [reply]
I'm not using CephFS for PVCs, so I don't think I would want to create a storage class for it, unless I'm missing something.  

The metadata pool is set to size 3, but it is also forcing the creation of a data pool that is also replicated.  So I end up with 3 pools for the filesystem - two replicated (metadata + data), and one erasure coded (data).

Ceph itself should be fine with only two of those three pools (replicated metadata and erasure coded data), at least from what I can tell from examples I've found online.

The extra pool shouldn't be a huge problem since I can always scale it down and leave it empty.  It just seems like an unnecessary constraint.

**Reply by travisn** [reply]
How are you using CephFS? The way to use it with Rook is to define a storage class and use the ceph-csi driver for cephfs PVs.

**Reply by rich0** [reply]
I'm using Rook to manage/provide CephFS, not consume it.  I mount it using the native k8s cephfs volume support (you just give it the monitor IPs).  I would love if Rook provided a way to mount CephFS without a PVC.  I haven't messed around with Rook CephFS PVCs but it looks like it would be pretty cumbersome to mount the same filesystem/path in multiple namespaces using it, and I also would need to use it outside of k8s as well.  

I am using the Rook CSI for PVCs that use RBD.

**Reply by travisn** [reply]
Rook only supports the CSI drivers. 

CephFS volumes can be used across namespaces with a [static PVC](https://github.com/ceph/ceph-csi/blob/devel/docs/static-pvc.md). 

For access outside K8s, [NFS](https://rook.io/docs/rook/latest/Storage-Configuration/NFS/nfs/) is the recommendation.

**Reply by rich0** [reply]
Hmm, I do have nodes that need access to the root, and I don't use subvolumes, but the static PVC might be useful (I'll have to experiment with giving it arbitrary paths).  If nothing else it means I won't be having to deal with the risk of changing monitor IPs.  Thanks for mentioning that.  It seems easier to use than what is in the docs, and it is nice that they allow specifying a namespace for the secret so that it doesn't have to be copied all over the place (which is necessary when using native k8s cephfs volumes).

I should just mark this as answered though.  My original problem was with directing writes to the appropriate pool and that has been solved.  The rest has been helpful all the same!
